pybullet_toolkit/robots/ur5_rg2.py

```diff
@@ -80,43 +86,92 @@ def place(self, pos, rot, offset, dynamic=True):
 
   def moveTo(self, pos, rot, dynamic=True):
     ''''''
-    ik_solve = pb.calculateInverseKinematics(self.id, self.end_effector_index, pos, rot)
-
-    if dynamic:
-      ee_pos = self._getEndEffectorPosition()
-      self._sendPositionCommand(ik_solve)
-      past_ee_pos = deque(maxlen=5)
-      while not np.allclose(ee_pos, pos, atol=0.01):
-        # time.sleep(0.005)
-        pb.stepSimulation()
-
-        # Check to see if the arm can't move any close to the desired position
-        if len(past_ee_pos) == 5 and np.allclose(past_ee_pos[0], past_ee_pos):
-          break
-
-        past_ee_pos.append(ee_pos)
-        ee_pos = self._getEndEffectorPosition()
-    else:
-      self._setJointPoses(ik_solve)
+    close_enough = False
+    it = 0
+    threshold = 1e-3
+    max_iteration = 100
+
+    while not close_enough and it < max_iteration:
+      ik_solve = pb.calculateInverseKinematics(self.id, self.end_effector_index, pos, rot)[:-2]
+      if dynamic:
+        self._sendPositionCommand(ik_solve)
+        past_joint_pos = deque(maxlen=5)
+        joint_state = pb.getJointStates(self.id, self.arm_joint_indices)
+        joint_pos = list(zip(*joint_state))[0]
+        while not np.allclose(joint_pos, ik_solve, atol=1e-2):
+          pb.stepSimulation()
+          # Check to see if the arm can't move any close to the desired joint position
+          if len(past_joint_pos) == 5 and np.allclose(past_joint_pos[-1], past_joint_pos, atol=1e-3):
+            break
+          past_joint_pos.append(joint_pos)
+          joint_state = pb.getJointStates(self.id, self.arm_joint_indices)
+          joint_pos = list(zip(*joint_state))[0]
+      else:
+        self._setJointPoses(ik_solve)
+
+      ls = pb.getLinkState(self.id, self.end_effector_index)
+      new_pos = list(ls[4])
+      new_rot = list(ls[5])
+      close_enough = np.allclose(np.array(new_pos + new_rot), np.array(list(pos) + list(rot)), atol=threshold)
+      it += 1
+    pass
+
+    # # ik_solve = pb.calculateInverseKinematics(self.id, self.end_effector_index, pos, rot)
```

Comment:
Might as well get rid of all these comments. If we need to look at this code later there is always the commit history.

Reply:
yeah that make sense.